File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
'''
    module defines a rectangle class
'''


class Rectangle:
    '''
        class that define a rectangle
    '''

    def __init__(self, width=0, height=0):
        self.width = width
        self.height = height

    @property
    def width(self):
        return self.__width

    @width.setter
    def width(self, value):
        if not isinstance(value, int):
            raise TypeError("width must an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value

    @property
    def height(self):
        return self.__height

    @height.setter
    def height(self, value):
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value

    def area(self):
        return self.height * self.width

    def perimeter(self):
        if self.height == 0 or self.width == 0:
            return 0
        else:
            return ((self.height * 2) + (self.width * 2))

    def __str__(self):
        string = ""
        if self.height == 0 or self.width == 0:
            return string
        else:
            for h in range(self.height):
                for w in range(self.width):
                    string += "#"
                string += "\n"
        return string[:-1]

    def __repr__(self):
        return "Rectangle(width={}, height={})".format(self.width, self.height)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    let str = '';

    for (let i = 0; i < this.height; i++) {
      for (let i = 0; i < this.width; i++) {
        str += 'X';
      }

      if (i !== this.height - 1) {
        str += '\n';
      }
    }
    if (str !== '') {
      console.log(str);
    }
  }

  rotate () {
    [this.width, this.height] = [this.height, this.width];
  }

  double () {
    [this.width, this.height] = [this.width * 2, this.height * 2];
  }
}

module.exports = Rectangle;
